pkg/controller/machine/machine_controller_phases.go

```diff
@@ -248,6 +249,22 @@ func (r *ReconcileMachine) reconcileInfrastructure(ctx context.Context, m *v1alp
 		return errors.Errorf("retrieved empty Spec.ProviderID from infrastructure provider for Machine %q in namespace %q", m.Name, m.Namespace)
 	}
 
+	addresses, _, err := unstructured.NestedSlice(infraConfig.Object, "status", "addresses")
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve addresses from infrastructure provider for Machine %q in namespace %q", m.Name, m.Namespace)
+	}
+
+	if len(addresses) > 0 {
+		addressesBytes, err := json.Marshal(addresses)
+		if err != nil {
+			return errors.Wrapf(err, "failed to encode addresses from infrastructure provider for Machine %q in namespace %q", m.Name, m.Namespace)
+		}
+
+		if err := json.Unmarshal(addressesBytes, &m.Status.Addresses); err != nil {
+			return errors.Wrapf(err, "failed to decode addresses from infrastructure provider for Machine %q in namespace %q", m.Name, m.Namespace)
+		}
+	}
```

Comment:
Is there a better way to do the conversion rather than roundtripping through json? If not, should this be a helper function since there will be additional places where we need to do this.

Reply:
You could try something like this
```go
type infra struct {
  Status infraStatus
}

type infraStatus struct {
  Addresses []corev1.NodeAddress
}

var i infra
runtime.FromUnstructured(infraConfig.Object, &i)
// check error
m.Status.Addresses = i.Status.Addresses
```

You may be able to go straight into m.Status.Addresses without needing the temporary structs (not sure though).

Reply:
@ncdc I can't find `runtime.FromUnstructured` in the current vendored version, I'll move the logic to the util package for now

Reply:
It must be in a different package. 1 sec.

Reply:
`runtime.DefaultUnstructuredConverter.FromUnstructured`

Reply:
I tested it out and it seems it requires a lot of manual work to get a single field. The json library is more flexible, ptal at the latest changes and let me know if that seems like a good approach

Reply:
There are helpers: `unstructured.NestedString` for example. Does that help?

Reply:
Following @detiber's feedback on making the json part outside, I added a generic function to deal with this https://github.com/kubernetes-sigs/cluster-api/pull/1174/files#diff-6c6a7fefe37261e2fa8a010cf9e7ae08R250